README.md

```diff
@@ -1,53 +1,251 @@
+
+Impagination is built using an event-driven immutable style, so it is
+ideal for use with UI frameworks like Ember, Angular, or React. That
+said, it has zero dependencies apart from JavaScript, so it can be
+used from node as well.
+
+## Usage
+
+To get started, create a dataset. There are only two required parameters
+`fetch`, and `observe`:
+
+```javascript
+import { Dataset } from 'impagination';
+
+let state = null;
+
+let dataset = new Dataset({
+  // how many records should we "keep ahead"? (default = pageSize)
+  loadHorizon: 10,
+  // fetch in pages of 5 (default 10)
+  pageSize: 5,
+  // this fake fetch function returns a page where the "records" *are*
+  // the offsets
+  fetch: function(pageOffset, pageSize, stats) {
+    stats.totalPages = 5;
+    return new Promise(function(resolve) {
+      resolve(return new Array(pageSize).fill(0).map(function(zero, i)) {
+        return pageOffset * pageSize + i;
+      });
+    });
+  },
+  //this function is invoked whenever a new state is generated.
+  observe: function(nextState) {
+    state = nextState;
+  }
+});
+```
+
+This will emit a state immediately, however this state will not have
+anything in it, but That's because we haven't told the dataset where we
+want to start reading from.
+
+```javascript
+state.length //=> 0;
+state.get(0) //=> null;
+```
+
+To tell where to start reading, you update the dataset's "read
+offset". It's the only "imperative" API that impagination exposes, and
+it indicates where you're interested in accessing records. Let's start
+at the beginning.
+
+```javascript
+dataset.setReadOffset(0);
+```
+
+Immediately, this will call fetch twice (for records 0-4, and 5-9),
+and emit a new state indicating that these records are in flight.
+
+```javascript
+state.length //=> 10
+let record = state.get(7)
+record.isPending //=> true
+record.isResolved //=> false
+record.content //=> null
+```
+
+### Load Horizon
+
+How did it know which records to fetch? The answer is in the
+`loadHorizon` parameter that we passed into the constructor. This
+tells the dataset, that it should keep all records within 10 of the
+current read offset loaded. That's why it fetched the first two
+pages. Now logicially, our dataset looks like this:
+
+> Note: `*` indicates that the record is pending.
+
+
+```
+           Read
+          Offset
+             ┃
+             ┃
+<──────────Load Horizon──────────>
+             ┃
+             ▼
+             ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐
+              * * * * * * * * * *
+             ◇ ─ ─ ─ ─ ◇─ ─ ─ ─  ┘
+             │         │
+             p0        p1
+```
+
+At some point, the request for the first page resolves. At that point,
+the dataset will emit a new state with the resolved records. That
+state will still contain the pending records as well as the freshly
+loaded.
+
+
+```javascript
+//get a record off the first page
+record = state.get(3);
+record.isResolved //=> true
+record.content //=> 3
```

Comment:
Or were you suggesting that we have our fetch function return an object instead of an integer. E.g.:

```js
{
  offset: 3,
  description: "This is the data at offset 3"
}
```

so then:

```js
record.content.description //=> This is the object at offset 3
```